### PR TITLE
[PYSPARK] [CONNECT] [ML] PySpark UDF supports python package dependencies

### PR DESCRIPTION
**Pull request by WeichenXu123**
#### Make the pyspark UDF support annotating python dependencies and when executing UDF, the UDF worker creates a new python environment with provided python dependencies.

 - Supported spark mode: spark connect model / legacy mode
 - Supported UDF type: All kinds of pyspark UDFs
    - SQL_BATCHED_UDF
    - SQL_SCALAR_PANDAS_UDF
    - SQL_GROUPED_MAP_PANDAS_UDF
    - SQL_GROUPED_AGG_PANDAS_UDF
    - SQL_WINDOW_AGG_PANDAS_UDF
    - SQL_SCALAR_PANDAS_ITER_UDF
    - SQL_MAP_PANDAS_ITER_UDF
    - SQL_COGROUPED_MAP_PANDAS_UDF
    - SQL_MAP_ARROW_ITER_UDF
    - SQL_GROUPED_MAP_PANDAS_UDF_WITH_STATE

#### Implementation sketch
 - Before starting the pyspark UDF worker process, the python environment with provided python packages is created, and pyspark UDF worker process is spawned using the provided python environment instead of default configured pyspark python. 
 - Using `virtualenv` to create a python environment based on current python environment that pyspark uses, then using `pip install` to install provided python packages.
 - If user configures a NFS directory that is accessible by all spark nodes (readable/writable to spark driver, readable to all spark workers), then it prepares the python environment in driver side, otherwise it creates the python environment in spark worker side.
 - The python environment is cached in spark driver or worker side (depending on NFS directory enabled or not), we uses SHA1 over sorted pip requirements list as the python environment caching key.

#### TODOs
 - For chained UDF functions, if they have different pip_requirements settings, we need to split them into separate PythonRunners
 - In frontend, the PR currently only supports annotating `pip_requirements` for `pandas_udf`, but for other types of UDFs, and for `mapInPandas` / `mapInArrow` the `pip_requirements` argument haven't been added.
 - Supports annotating python version for pyspark UDF, and in UDF execution, downloading python using provided python version and creating python environment using provided python version.
 - Using file lock during python environment creation, to avoid race conditions.
 - Unit tests

### Why are the changes needed?

 - For spark connect case, the client python environment is very likely to be different with pyspark server side python environment, this causes user's UDF function execution failure in pyspark server side.
 - Some machine learning third-party library (e.g. MLflow) requires pyspark UDF supporting  dependencies, because in ML cases, we need to run model inference by pyspark UDF in the exactly the same python environment that trains the model. Currently MLflow supports it by creating a child python process in pyspark UDF worker, and redirecting all UDF input data to the child python process to run model inference, this way it causes significant overhead, if pyspark UDF support builtin python dependency management then we don't need such poorly performing approach.

### Does this PR introduce _any_ user-facing change?
```

@pandas_udf("string", pip_requirements=...)

```

`pip_requirements` argument means either an iterable of pip requirement strings (e.g. ``["scikit-learn", "-r /path/to/req2.txt", "-c /path/to/constraints.txt"]``) or the string path to a pip requirements file path on the local filesystem (e.g. ``"/path/to/requirements.txt"``) represents the pip requirements for the python UDF.

### How was this patch tested?

Unit tests to be added.

Manually tests:

```
import pandas as pd
from pyspark.sql.functions import pandas_udf

sc.setLogLevel("INFO")

@pandas_udf("string", pip_requirements=["PyYAML==6.0"])
def to_upper(s: pd.Series) -> pd.Series:
    import yaml
    return s.str.upper() + f"yaml-version: {yaml.__version__}"

df = spark.createDataFrame([("John Doe",)], ("name",))
df.select(to_upper("name")).show(truncate=False)
```

Run above code in spark legacy mode or spark connect mode.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/api/python/PythonEnvSetup.scala [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    val headPipDeps = funcs.head.funcs.head.pipDependencies
+    val headPipConstraints = funcs.head.funcs.head.pipConstraints
+
+    for (chainedFunc <- funcs) {
+      for (simpleFunc <- chainedFunc.funcs) {
+        if (simpleFunc.pipDependencies != headPipDeps ||
+          simpleFunc.pipConstraints != headPipConstraints
+        ) {
+          // TODO: For this case, we should split current python runner
+          //  into multiple python runners.
```
Comment (WeichenXu123) [inline core/src/main/scala/org/apache/spark/api/python/PythonEnvSetup.scala:100]: CC @HyukjinKwon

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [hunk 2]
```diff
@@ -333,6 +333,11 @@ message PythonUDF {
   bytes command = 3;
   // (Required) Python version being used in the client.
   string python_ver = 4;
+  // (Optional) Python function pip dependencies
```
Comment (zhengruifeng) [inline connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:336]: I think we may add a new Class and proto message for dependency

File: core/src/main/scala/org/apache/spark/api/python/PythonEnvSetup.scala (hunk 1 above)
Comment (zhengruifeng) [reply core/src/main/scala/org/apache/spark/api/python/PythonEnvSetup.scala:100]: maybe we could first detect whether the two envs can be merged, e.g.:
```
env1:
numpy

env2:
numpy
scipy
```

File: python/pyspark/pythonenv/pip_req_parser.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        raise ValueError(f"This pip requirement line is invalid: '{line}'")
+    if match.group(1).lower() == "pyspark":
+        raise ValueError(
+            "pyspark dependency or constraint is disallowed to set for pyspark UDF, "
+            "pyspark UDF worker python environment always install pyspark package with "
+            "exactly the same version with server side spark."
+        )
+
+
+def _parse_requirements(requirements, is_constraint, base_dir=None):
```
Comment (zhengruifeng) [inline python/pyspark/pythonenv/pip_req_parser.py:98]: is it possible to reuse `pip` instead?

`from pip._internal.req import parse_requirements`

File: core/src/main/scala/org/apache/spark/api/python/PythonEnvSetup.scala (hunk 1 above)
Comment (WeichenXu123) [reply core/src/main/scala/org/apache/spark/api/python/PythonEnvSetup.scala:100]: Does this TODO conflict with https://github.com/databricks/runtime/pull/55522 ?

File: python/pyspark/pythonenv/pip_req_parser.py (hunk 3 above)
Comment (WeichenXu123) [reply python/pyspark/pythonenv/pip_req_parser.py:98]: I feel it is unsafe to call `pip._internal` that is not public interface

File: core/src/main/scala/org/apache/spark/api/python/PythonEnvSetup.scala (hunk 1 above)
Comment (WeichenXu123) [reply core/src/main/scala/org/apache/spark/api/python/PythonEnvSetup.scala:100]: Btw, this TODO case is a minor case, usually UDFs in one spark job only requires one kind of python dependency settings.

**Comment by zhengruifeng** [comment]
also cc @cloud-fan @hvanhovell

File: core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala [hunk 4]
```diff
@@ -160,8 +164,20 @@ private[spark] abstract class BasePythonRunner[IN, OUT](
       envVars.put("PYTHON_FAULTHANDLER_DIR", BasePythonRunner.faultHandlerLogDir.toString)
     }
 
+    // If local property "pythonEnv.allSparkNodesReadableRootEnvDir" is set,
+    // then driver side has initiated a python environment in this directory.
+    // `env.createPythonWorker` can reuse the python environment set up in driver side.
+    val rootPythonEvnDir = Option(
+      context.getLocalProperty("pythonEnv.allSparkNodesReadableRootEnvDir")
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala:171]: Can we start the name with `spark`.

File: core/src/main/scala/org/apache/spark/api/python/PythonEnvManager.scala [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    //    concurrent creation/deletion causes race condition
+    // TODO: use file lock instead to ensure safety, e.g. the case that multiple
+    //  spark executors running on the same machine.
+    synchronized {
+      val key = getPythonEnvironmentKey(pipDependencies, pipConstraints)
+      val envDir = Path.of(rootEnvDir, key).toString
+
+      if (!new File(envDir).exists()) {
+        try {
+          createPythonEnvironment(pythonExec, rootEnvDir, envDir, pipDependencies, pipConstraints)
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/PythonEnvManager.scala:79]: When do we remove this new env?

**Pull request closed without merging**